server/src/main/java/org/candlepin/controller/RevocationOp.java

```diff
@@ -0,0 +1,171 @@
+        // We have to wait until we get here so that share pool entitlements we want revoked are gone
+        for (Entitlement entitlement : entitlementsToAdjust.keySet()) {
+            try {
+                poolManager.adjustEntitlementQuantity(entitlement.getConsumer(),
+                    entitlement, entitlementsToAdjust.get(entitlement).intValue());
+            }
+            catch (EntitlementRefusedException e) {
+                // TODO: Could be multiple errors, but we'll just report the first one for now:
+                throw new ForbiddenException(e.getResults().values().iterator().next().getErrors()
+                        .get(0).toString());
+            }
+        }
+
+    }
+
+    /**
+     * The first part of the adjustment is to reduce the shared pool source entitlements for counts that are not
+     *  in use as entitlements from the shared pool.
+     *
+     * @param pool
+     * @return boolean
+     */
+    private boolean reduceSharedPools(Pool pool) {
+        newConsumed = pool.getConsumed();
+        long over = pool.getConsumed() - pool.getQuantity();
+        for (Pool sPool : sharedPools.get(pool)) {
+            if (over > 0) {
+                long excessCount = sPool.getQuantity() - sPool.getConsumed();
+                long newQuantity = 0;
+                if (excessCount > 0) {
+                    Entitlement shareEnt = sPool.getSourceEntitlement();
+                    if (over >= excessCount) {
+                        newQuantity = sPool.getQuantity() - excessCount;
+                        over = over - excessCount;
+                        newConsumed -= excessCount;
+                    }
+                    else {
+                        newQuantity = shareEnt.getQuantity() - over;
+                        newConsumed -= over;
+                        over = 0L;
+                    }
+                    addEntitlementToAdjust(shareEnt, newQuantity);
+                }
+            }
+        }
+        return over > 0;
+    }
+
+    /**
+     * In the second part, the list of entitlements is compiled from the main pool plus the shared pools
+     * It is sorted in the order of LIFO. Entitlements are put in the revoke list until the count is
+     * acceptable for the main pool. Any entitlements that came from a shared pool are also reflected in
+     * the adjustment for the source entitlement for that shared pool.
+     *
+     * @param pool
+     */
+    private void determineExcessEntitlements(Pool pool) {
+        List<Pool> pools = new ArrayList<Pool>();
+        pools.add(pool);
+        pools.addAll(sharedPools.get(pool));
+        List<Entitlement> freeEntitlements = this.poolCurator.retrieveOrderedEntitlementsOf(pools);
+
+        long existing = pool.getQuantity();
+        for (Entitlement ent : freeEntitlements) {
+            if (newConsumed > existing) {
+                if (!ent.getConsumer().isShare()) {
```

Comment:
in that case, before we revoke ents, we should go through the ents to adjust and if any of them are to be set to 0, we should revoke it instead. that we dont leave behind ents with quantity 0.